src/Caching/Hybrid/src/Internal/BufferChunk.cs

```diff
@@ -0,0 +1,77 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.Caching.Hybrid.Internal;
+
+// used to convey buffer status; like ArraySegment<byte>, but Offset is always
+// zero, and we use the MSB of the length to track whether or not to recycle this value
+internal readonly struct BufferChunk
+{
+    private const int MSB = (1 << 31);
+
+    private readonly int _lengthAndPoolFlag;
+    public byte[]? Array { get; } // null for default
+
+    public int Length => _lengthAndPoolFlag & ~MSB;
+
+    public bool ReturnToPool => (_lengthAndPoolFlag & MSB) != 0;
+
+    public byte[] ToArray()
+    {
+        var length = Length;
+        if (length == 0)
+        {
+            return [];
+        }
+
+        var copy = new byte[length];
+        Buffer.BlockCopy(Array!, 0, copy, 0, length);
+        return copy;
+    }
+
+    public BufferChunk(byte[] array)
+    {
+        Debug.Assert(array is not null, "expected valid array input");
+        Array = array;
+        _lengthAndPoolFlag = array.Length;
+        // assume not pooled, if exact-sized
+        Debug.Assert(!ReturnToPool, "do not return right-sized arrays");
```

Comment:
By right-sized, I mean this is a byte[] without a length - which means it came from the legacy byte[] GetValue API, and is not a pooled buffer. We're simply asserting "we don't expect to recycle this, because it isn't a pool buffer"

Reply:
Sorry, I'm going to push since this seems important and I'm still not following it.  Are you saying that the built-in `byte[].Length` will return a negative value if it came from the pool?  Did we forcibly set the sign bit somewhere?  Or maybe there's some reason this isn't the built-in `byte[].Length`?

Reply:
Yes (semantically, at least) and yes; when passing just a byte[], we have to assume that we mean all of it, i.e. that it is right-sized rather than an oversized buffer with an active portion